ritasamir/assignment2/Q_02.java

```diff
@@ -0,0 +1,55 @@
+package assignment_2;
+
+import java.util.ArrayDeque;
+import java.util.Queue;
+
+public class Q_02 {
+	Q_01 ancestors;
+
+	/**
+	 * search the binary tree for
+	 * 
+	 * @param head
+	 *            the root of the binary tree
+	 * @param key1
+	 * @param key2
+	 * @return node which represent the common ancestor for both key1 and key2
+	 */
+	public Node getCommonAncestor(Node head, int key1, int key2) {
+		ancestors = new Q_01();
+		Node node = ancestors.searchTree(head, key1);
+		while (node.getParent() != null) {
+			node = node.getParent();
+			if (isChildOf(node, key2)) {

```

Comment:
Using a queue in the isChildOf is still using extra memory, which is not allowed for this assignment. :(
How about implementing this method as something among the lines of the solution we've discussed on Friday:

```
public Node getCommonAncestor(Node head, int key1, int key2) {
    Node node1 = searchTree(head, key1);
    Node node2 = searchTree(head, key2);

    return getCommonAncestor(node1, node2);
}

private Node getCommonAncestor(Node node1, Node node2) {
    // if any of the nodes is null, there is no common ancestor
    if (node1 == null || node2 == null) {
        return null;
    }

    // check if we already have a common ancestor (ourselves)
    if (node1 == node2) {
        return node1;
    }

    // go from node2 through its parents to check if we find the common ancestor
    Node parent = node2.parent;
    while (parent != null) {
        if (parent == node1) {
            // found it!
            return parent;
        }
        parent = parent.parent;
    }

    // we didn't have node2 in node1's subtree, check node1's ancestors recursively
    return getCommonAncestor(node1.parent, node2);
}
```

Reply:
i edited my code in a way that it doesn't use extra memory , please review it and tell me if i need to change or edit it